### ISSUE TITLE
Distributed inference

### ISSUE THREAD

**Issue by isaacleeai** [facebookresearch/flashlight#29]
1. Is there a way to enable distributed inference, instead of training? I see that when you use `eval()` method on `fl::sequence`, it turns gradient off. So is running "training with gradient off" computationally equivalent to inference? If this is the case, then I guess there is no need to separately implement distributed inference, but if it is not, then I would love to know where to get started for my own implementation.

2. is it possible to distribute the work ( both training and inference ) across multiple servers each with multiple GPUs, or does it only work for a single server with multiple GPU? If any of these features are missing, will they be coming out soon?

3. Lastly, if distributed inference is not supported, what would be the recommended environment / library to enable distributed inference on multiple servers each with multiple GPUs?

**Comment by jacobkahn** [comment]
@isaacleeai — to answer your questions:
1. You're correct; there's no explicit support for distributed inference right now. The only synchronization that needs to happen during training is with initial model parameters and gradients.
- flashlight implements a [data parallel](https://en.wikipedia.org/wiki/Data_parallelism) distributed model. This means that in a distributed environment, each GPU has the full model and all parameters, and executes `forward` and `backward` independently; samples are divided up between GPUs. Since, at inference time, you have the entire trained model, there's no need to synchronize across GPUs. Giving each GPU the model and having it perform inference independently over some samples should be straightforward to implement using the existing dataset tools.

2. You can distributed training across multiple GPUs, and also across multiple nodes each with multiple GPUs — this functionality already exists ([see the distributed docs](https://fl.readthedocs.io/en/latest/distributed.html)). MPI will effectively communicate between nodes. We've used flashlight to train models with over 64 GPUs per-job.

3. Your environment doesn't really matter in terms of GPU interconnect/PCIe topology, since, again, there shouldn't be a need for any reduction operations between GPUs during inference.

**Comment by isaacleeai** [comment]
1. Is the data parallel distributed model what is commonly used in the industry? It seems like storing the entire model on each device seems potentially wasteful of DRAM.

**Comment by jacobkahn** [comment]
Data parallel training with minibatches is pretty common. It's the default for Tensorflow PyTorch, MxNet, etc. Even models with many parameters aren't too expensive to replicate. Model parallelism is also supported by some frameworks, but it usually involves rolling things yourself/hand-picking which parts of the computation graph are on which machines.

**Comment by isaacleeai** [comment]
This seems like a more general question, but if you wouldn't mind me asking: 
1. what type of parallel system is used by Facebook ( or any other industry giants ) to train models?
2. how about for inference ( i.e. production )?

Thanks a lot for all your helpful answers.

**Comment by jacobkahn** [comment]
1. Data-parallel training is very common.
2. This is highly variable and depends on the architecture, task, etc. Typically, inference is simpler and no distribution is needed.
3. I don't quite understand your question — I'd take a look at [these dependencies](https://fl.readthedocs.io/en/latest/installation.html#distributed-training-dependencies) and [these docs](https://fl.readthedocs.io/en/latest/distributed.html). With the dependencies listed, you can run multi-node distributed training.
